flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/TemporalJoinTestPrograms.java

```diff
@@ -84,6 +130,36 @@ public class TemporalJoinTestPrograms {
                                     + "ON o.currency = r.currency ")
                     .build();
 
+    static final TableTestProgram TEMPORAL_JOIN_TABLE_JOIN_NESTED_KEY =
+            TableTestProgram.of(
+                            "temporal-join-table-join-nested-key",
+                            "validates temporal join with a table when the join keys comes from a nested row")
+                    .setupTableSource(ORDERS_WITH_NESTED_ID)
+                    .setupTableSource(RATES)
+                    .setupTableSink(AMOUNTS)
+                    .runSql(
+                            "INSERT INTO MySink "
+                                    + "SELECT amount * r.rate "
+                                    + "FROM OrdersNestedId AS o "
+                                    + "JOIN RatesHistory FOR SYSTEM_TIME AS OF o.rowtime AS r "
+                                    + "ON o.nested_row.currency = r.currency ")
+                    .build();
+
+    static final TableTestProgram TEMPORAL_JOIN_TABLE_JOIN_KEY_FROM_MAP =
+            TableTestProgram.of(
+                            "temporal-join-table-join-key-from-map",
+                            "validates temporal join with a table when the join key comes from a map value")
+                    .setupTableSource(ORDERS_WITH_NESTED_ID)
+                    .setupTableSource(RATES)
+                    .setupTableSink(AMOUNTS)
+                    .runSql(
+                            "INSERT INTO MySink "
+                                    + "SELECT amount * r.rate "
+                                    + "FROM OrdersNestedId AS o "
+                                    + "JOIN RatesHistory FOR SYSTEM_TIME AS OF o.rowtime AS r "
+                                    + "ON o.nested_map['currency'] = r.currency ")
```

Comment:
Tests are added for covering what happens with the left-hand side of the join.  

Do we need to verify anything for the right-hand side and/or both sides?

Also, can join on the output of a function call?  (E.g., `ON UPPERCASE(o.nested_map['currency']) = r.currency`?)

Reply:
I tried adding a case with a projection on the ride side of the input.

The change introduced here works fine, unfortunately the query fails with a different condition unmet. The temporal table join requires the right side of the join (the versioned table) to be a `PRIMARY KEY`. `PRIMARY KEY` can only be defined on a source table, projections remove that constraint. That means the change improves the situation only for the left side of the join.

I'll still add a case with a function on the left side.